Improve i3.nix to handle options as list like in sway, adjusted functions for less new-lines

SebTM opened the pull request:
### Description
Issue #2030 should describe it well :)
Didn't fixed the tests yet, just wanted to ask if there is feedback before about the way2go here? :)

### Checklist
- [X] Change is backwards compatible.
- [X] Code formatted with `./format`.
Yes but formats some code parts really bad?

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

sumnerevans commented on modules/services/window-managers/i3-sway/lib/functions.nix:
```diff
@@ -77,88 +77,70 @@ rec {
-          optionalString (statusCommand != null)
-          "status_command ${statusCommand}"
+          concatStringsSep "\n" (notEmpty [
+            "${optionalString (id != null) "id ${id}"}"
```
There is no need to wrap any of these `optionalString`s in another string:

```suggestion
            (optionalString (id != null) "id ${id}")
```

SebTM replied:
Thanks for the feedback, will adjust and fix the tests just wanted to get sure that I'm not on the wrong path...
Would it be useful to have something like not Empty in lib or is there something similar?

sumnerevans replied:
I think it makes sense to move `notEmpty` to lib so that we can take advantage of it across the codebase. I don't know of anything like that in nixpkgs lib or hm lib.

SebTM replied:
@sumnerevans ping

SebTM replied:
@sumnerevans bump?

sumnerevans reviewed (changes requested):
Seems like the right direction. Please update the tests to make sure that you didn't introduce any regressions.

SebTM commented:
Fixed the tests, format and adjusted the writeText in i3/sway to follow this approach. Would be nice if you could review it again :)

berbiche commented on modules/services/window-managers/i3-sway/lib/functions.nix:
```diff
@@ -77,88 +77,67 @@ rec {
-    ${optionalString (right != null) "gaps right ${toString right}"}
-
-    ${optionalString smartGaps "smart_gaps on"}
-    ${optionalString (smartBorders != "off") "smart_borders ${smartBorders}"}
+    ${concatStringsSep "\n" (notEmpty [
+      (optionalString (inner != null) "gaps inner ${toString inner}")
+      (optionalString (outer != null) "gaps outer ${toString outer}")
+      (optionalString (horizontal != null)
+        "gaps horizontal ${toString horizontal}")
+      (optionalString (vertical != null) "gaps vertical ${toString vertical}")
+      (optionalString (top != null) "gaps top ${toString top}")
+      (optionalString (bottom != null) "gaps bottom ${toString bottom}")
+      (optionalString (left != null) "gaps left ${toString left}")
+      (optionalString (right != null) "gaps right ${toString right}")
+      (optionalString smartGaps "smart_gaps on")
+      (optionalString (smartBorders != "off") "smart_borders ${smartBorders}")
+    ])}
   '';
```
Can you remove the outer string? It's not needed anymore.

SebTM replied:
With outer string you mean `${concatStringsSep "\n" (notEmpty [`?

SebTM replied:
Why are they not needed anymore? I can't find a drop notice/some explanation at i3-gaps...

berbiche commented on modules/services/window-managers/i3-sway/lib/functions.nix:
```diff
@@ -176,4 +155,7 @@ rec {
     "for_window ${criteriaStr criteria} ${command}";
   workspaceOutputStr = item:
     ''workspace "${item.workspace}" output ${item.output}'';
+
+  # Source: https://github.com/nix-community/home-manager/blob/master/modules/services/xidlehook.nix#L10 | Optional strings will else generate empty lines
+  notEmpty = list: filter (x: x != "" && x != null) (flatten list);
```
Can this be moved before its usage?

SebTM replied:
If you mean moving it to lib yeah I was not sure if I have to split it up: https://github.com/nix-community/home-manager/pull/2314#discussion_r725502626

berbiche commented on modules/services/window-managers/i3-sway/i3.nix:
```diff
@@ -143,50 +143,47 @@ let
+        ++ mapAttrsToList assignStr assigns ++ map barStr bars
+        ++ optional (gaps != null) gapsStr
+        ++ map floatingCriteriaStr floating.criteria
+        ++ map windowCommandsStr window.commands ++ map startupEntryStr startup
```
```suggestion
        ++ map windowCommandsStr window.commands
        ++ map startupEntryStr startup
```

Will the formatter approve of this change?

berbiche commented on modules/services/window-managers/i3-sway/i3.nix:
```diff
@@ -143,50 +143,47 @@ let
+        (keybindingsStr { keybindings = keybindingsRest; })
+        (keycodebindingsStr keycodebindings)
+      ] ++ mapAttrsToList (modeStr false) modes
+        ++ mapAttrsToList assignStr assigns ++ map barStr bars
```
```suggestion
        ++ mapAttrsToList assignStr assigns
        ++ map barStr bars
```

berbiche reviewed (changes requested):
Looks good overall.
Would be cool if the indentation could also be preserved.

SebTM commented:
> Looks good overall. Would be cool if the indentation could also be preserved.

Tried my best in the new commits :)

SebTM commented:
Changed everything according feedback, one conversation left because of some things to check back :)
Kudo's to @Ma27 for supporting 🥇

berbiche commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -264,40 +264,37 @@ let
 
   configFile = pkgs.writeText "sway.conf" ((if cfg.config != null then
     with cfg.config; ''
```
Same here.

berbiche commented on modules/services/window-managers/i3-sway/lib/functions.nix:
```diff
@@ -77,88 +78,69 @@ rec {
       }
     '';
 
   gapsStr = with cfg.config.gaps; ''
```
Ditto about the unneeded string interpolation (`"${concatStringsSep}"`).

berbiche commented on modules/services/window-managers/i3-sway/lib/functions.nix:
```diff
@@ -20,9 +20,9 @@ rec {
     cfg.config.keybindings;
 
-  keybindingsStr = { keybindings, bindsymArgs ? "" }:
+  keybindingsStr = { keybindings, bindsymArgs ? "", prefix ? "" }:
```
Nice :+1:

sumnerevans replied:
Maybe call this `indent` instead of `prefix`?

berbiche reviewed (dismissed):
Could you move the lists.nix changes to a different PR and undo the changes in this PR?

SebTM commented:
> Could you move the lists.nix changes to a different PR and undo the changes in this PR?

Done, also did my best for the other points of your feedback :) This PR now depends on #2566 where the new lib-function is added. Would appreciate your re-review 👍🏻

SebTM commented:
Rebased on current master, Cleaned commit :)

sumnerevans commented:
Looks like the build is failing for some reason. Can you look into that?

sumnerevans commented on modules/services/window-managers/i3-sway/i3.nix:
```diff
@@ -143,51 +143,46 @@ let
+  startupEntryStr = { command, always, notification, workspace, ... }:
+    concatStringsSep " " [
+      (if always then "exec_always" else "exec")
+      (if (notification && workspace == null) then "" else " --no-startup-id")
```
I don't think that space is necessary since inside of a `concatStringsSep`, right?
```suggestion
      (if (notification && workspace == null) then "" else "--no-startup-id")
```

sumnerevans commented on modules/services/window-managers/i3-sway/i3.nix:
```diff
@@ -143,51 +143,46 @@ let
+      "floating_modifier ${floating.modifier}"
+      (windowBorderString window floating)
+      "hide_edge_borders ${window.hideEdgeBorders}"
+      "force_focus_wrapping ${if focus.forceWrapping then "yes" else "no"}"
```
Maybe it's worth extracting this out to a function while you are at it. See the `yesno` function in the neomutt module.

We could probably extract it out to lib as well, but I'm fine with doing that in a separate PR.

SebTM replied:
Created a follow up #2772 for it, as it would be done in a another PR anyway I guess...

sumnerevans commented on tests/modules/services/window-managers/i3/i3-bar-focused-colors-expected.conf:
```diff
@@ -98,13 +94,6 @@ bar {
     binding_mode #2f343a #900000 #ffffff
-  }
-  
 }
```
since you are already dealing with indents, this end `}` for the colors block should be indented. If too hard, it's not a big deal.

sumnerevans reviewed (changes requested):
Looks good, overall.

SebTM commented:
Rebased to current master, applied changes from feedback, fixed tests and found  `lists.subtractLists` which does exactly what I needed I guess :)

Tests seem to fail because of the issue in #2767 with PR #2770 pending ✌🏻

SebTM commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -263,53 +263,49 @@ let
+      ++ map windowCommandsStr window.commands # window commands
+      ++ map startupEntryStr startup # startup
+      ++ map workspaceOutputStr workspaceOutputAssign # custom mapping
+      ++ [ "\n" ])
```
This is a bit hacky, but otherwise exec is on the same line as the last workspaceOutputAssign... Maybe you have a better solution?

sumnerevans replied:
You could try moving the `concatStringsSep` above the `if` statement to encompass both cases.

SebTM replied:
Thanks you for the hint, it brought me on the right direction I think - I like the new solution a lot more ;)

sumnerevans reviewed (approved):
Sorry it's taken so long to look over this. I think it looks fine. If you want to try and refactor to avoid the `++ [ "\n" ]` then feel free to and I'll have another look.

SebTM commented:
Are the failing tests related? I would say no...

> herbstluftwm-simple-config: FAILED
> Expected /nix/store/hjxvnf979k037q74digq2k64w49z2n69-nmt-report-herbstluftwm-simple-config/normalized/autostart to be same as /nix/store/ci79wirs2xwxb23nfbd3cd14b0g4bzpv-autostart but were different:
> --- actual
> +++ expected
> @@ -1,4 +1,4 @@
> -#!/nix/store/00000000000000000000000000000000-bash-5.1-p16/bin/bash
> +#!/nix/store/00000000000000000000000000000000-bash-5.1-p12/bin/bash
>  shopt -s expand_aliases
>  
>  # shellcheck disable=SC2142
> For further reference please introspect /nix/store/hjxvnf979k037q74digq2k64w49z2n69-nmt-report-herbstluftwm-simple-config
> 
> mpd-basic-configuration: FAILED
> Expected /nix/store/pgj4vjhihn3vsx4n9865w4ik3vaqklkf-nmt-report-mpd-basic-configuration/normalized/mpd.service to be same as /nix/store/6ix31n8jxwzq5rzmyk0vdh8jkhs65y9j-basic-configuration.service but were different:
> --- actual
> +++ expected
> @@ -4,7 +4,7 @@
>  [Service]
>  Environment=PATH=/home/hm-user/.nix-profile/bin
>  ExecStart=@mpd@/bin/mpd --no-daemon /nix/store/00000000000000000000000000000000-mpd.conf
> -ExecStartPre=/nix/store/00000000000000000000000000000000-bash-5.1-p16/bin/bash -c "/nix/store/00000000000000000000000000000000-coreutils-9.0/bin/mkdir -p '/home/hm-user/.local/share/mpd' '/home/hm-user/.local/share/mpd/playlists'"
> +ExecStartPre=/nix/store/00000000000000000000000000000000-bash-5.1-p12/bin/bash -c "/nix/store/00000000000000000000000000000000-coreutils-9.0/bin/mkdir -p '/home/hm-user/.local/share/mpd' '/home/hm-user/.local/share/mpd/playlists'"
>  Type=notify
>  
>  [Unit]
> For further reference please introspect /nix/store/pgj4vjhihn3vsx4n9865w4ik3vaqklkf-nmt-report-mpd-basic-configuration

I've also added myself to CodeOwners for notifications on related things, if it's OK for your?

SebTM commented:
Rebased (now including fixed tests) and added commit as mention in #2815 :)
